spec/models/saved_page_spec.rb

```diff
@@ -1,4 +1,55 @@
 RSpec.describe SavedPage do
+  describe "validations" do
+    let(:saved_page) { FactoryBot.build(:saved_page) }
+
+    it "rejects duplicate saved pages, scoped to a user" do
+      saved_page.save!
+      duplicate = saved_page.dup
+      expect(duplicate).not_to be_valid
+      expect(duplicate.errors[:page_path]).to include("has already been taken")
+    end
+
+    it "allows duplicate saved pages, between different users" do
+      saved_page.save!
+      not_a_duplicate = FactoryBot.build(:saved_page, page_path: saved_page.page_path)
+      expect(not_a_duplicate).to be_valid
+    end
+
+    it "rejects nil" do
+      saved_page.page_path = nil
+      expect(saved_page).not_to be_valid
+      expect(saved_page.errors[:page_path]).to include("can't be blank")
+    end
+
+    it "rejects page paths that are not url safe string" do
+      saved_page.page_path = "/foo/bar"
+      expect(saved_page).not_to be_valid
+      expect(saved_page.errors[:page_path]).to include("must be a URL safe string")
+    end
+
+    it "allows saving URL safe strings" do
+      saved_page.page_path = CGI.escape("/foo/bar")
+      expect(saved_page).to be_valid
+    end
+
+    it "does not allow empty strings as paths" do
+      saved_page.page_path = ""
+      expect(saved_page).not_to be_valid
+    end
```

Comment:
Nitpick, feel free to ignore.
For consistency with other tests, this should come right after the `"rejects nil"` test, have title `"reject empty strings"` and have the error message expectation.

Reply:
Or the two could be tested together (since it's the same validation):
```ruby
it "rejects blank values as page paths" do
  [nil, ""].each do |blank_value|
    saved_page.page_path = blank_value
    expect(saved_page).not_to be_valid
    expect(saved_page.errors[:page_path]).to include("can't be blank")
  end
end
```